Add preference to change severity for missing project specific encoding

ingomohr opened the pull request:
Introduces a new preference that enables to specify if and using which
severity markers shall be added in case of missing project-specific
encoding settings.

This is the PR of the fork for issue [#166](https://github.com/eclipse-platform/eclipse.platform.resources/issues/166)

### Todo
* @merks I could use some help/hints w/ further unit tests.
** i.e. where am I supposed to integrate tests for this? Can I add an extra test class just for this?
* @merks Maybe you have an idea for this: I cannot seem to delete the "project encoding" markers.
** If I delete them, the pref change listener re-creates them right away, so that to the user this has the effect of "not deleted at all"
*** I checked the event type, the build kind and the delta type - but they're all exactly the same for both cases:
**** I delete the marker from the Problems view
**** I re-build (or clean and rebuild) the projects on the workspace
*** What I want is to be able to delete the markers - and then have a project-rebuild re-add the markers again.
*** Turning off "on access refresh" had no effect here.
*** I'm out of ideas atm.

### Here's what I did
#### Change
* Added the preference with a default value (which is WARN (as it was the severity before)
* Added a pref change listener as a service to remove/add/update the markers accordingly
* Added a severity provider class b/c I need access to the severity in both
** the new pref change listener I added and
** the existing `ValidateProjectEncoding` class
* Tested that provider using Mockito (which I added to the dependencies of the resources test fragment)
** Since I cannot mock a static constant (static methods can be mocked w/ Mockito, but constants apparently can't), I decided to use the same package for the test class so that I can override the "package-protected" getter for the InstanceScope.

#### Manual Integration Test
* Tested w/ my fork of [eclipse.platform.ui](https://github.com/ingomohr/eclipse.platform.ui)
* Modifying the severity (except "ignore") updated the markers as expected
* Modifying the severity to "ignore" removed the markers as expected
* Modifying the severity from "ignore" to something else added the markers were needed

Thank you in advance for having a look at the PR!

Kind regards
Ingo

merks commented:
It generally looks good. The corresponding nature test is here:

org.eclipse.core.tests.resources.NatureTest

I wonder if the ProjectEncodingPreferenceSeverityProvider  is a bit overkill because I don't see a corresponding thing for the Nature checking; I think it's using package protected helper methods.

I'm not sure I understand the problems with deleting the markers.  is the behavior you see different from deleting the marker for a missing nature?

ingomohr commented:
### Change After 1st Review
* Added unit tests (they all passed on the local setup; CI will give final answer)
* Removed ``ProjectEncodingPreferenceSeverityProvider`` and used static-"package protected"-method approach (instead)
** (consistency w/ the other impl)

merks commented:
Note that the tests that are failing all appear to pass on master:

https://ci.eclipse.org/platform/job/eclipse.platform.resources/job/master/lastBuild/testReport/org.eclipse.core.tests.resources/

ingomohr commented:
> Note that the tests that are failing all appear to pass on master:
> 
> https://ci.eclipse.org/platform/job/eclipse.platform.resources/job/master/lastBuild/testReport/org.eclipse.core.tests.resources/

Good morning, @merks . I'm on it. Those regression tests still boggle my mind atm, but this should be resolvable. :)

ingomohr commented:
#### Update After Fixing Failing Regression Tests
* The problem was that my preference change listener actually also did things when the resource changed
  * That, however, was already the job of ``ValidateProjectEncoding``.
  * After removing the unnecessary code CI went "green".
* Problem is now:
  * My pref change listener technically still is an ``IResourceChangeListener`` simply because I cannot register ``IPreferenceChangeListener``s via OSGi service.
  * I could register it at the activator - but instead I would like to add support for adding pref change listeners via service, too. However, I don't want to mix that into this PR.

What do you think, @merks ?

merks commented:
@ingomohr  Good question.   I assume you'd need to listen to this one node in particular so the service would need to provide a fair bit of information to get the right preference node:
```
InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES);
```
@laeubi is likely to have some insights/thoughts about registering a preference listener as a service.

This area in the code accesses the instance preferences so that's probably the appropriate place to register a listener programmatically:

https://github.com/eclipse-platform/eclipse.platform.resources/blob/625a25fa7752eb3cfe33216d314ad82501a38c74/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java#L2242-L2246

laeubi commented:
> * Problem is now:

I'm not sure if I fully understand *what* the problem is, but here is an example registering the listeners via DS:

https://github.com/eclipse-platform/eclipse.platform.resources/blob/master/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/CheckMissingNaturesListener.java

ingomohr commented:
> > * Problem is now:
> 
> I'm not sure if I fully understand _what_ the problem is, but here is an example registering the listeners via DS:
> 
> https://github.com/eclipse-platform/eclipse.platform.resources/blob/master/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/CheckMissingNaturesListener.java

Hi @laeubi  Thanks for the link. The problem is that the service accepts ``IResourceChangeListener``s. What I want to register, however, is not an ``IResourceChangeListener`` but an ``IPreferenceChangeListener``.
* At the moment, I add that listener via service as resource change listener and don't implement the ``resourceChanged``method - but that feels quite "hacky". :)

So, what we imho need to decide (for the ticket #166 to which this PR refers):
* Do we keep the impl the way it is and create another issue to add the feature to add ``IPreferenceChangeListener``s via service, too
* or do I register the pref change listener for this PR at the Activator class instead ( ``ResourcesPlugin``)?
* or do we do something else?

merks commented:
My 2 cents...  Programmatically register to get this PR to closure.  Then, if someone wants to work on being able to register a preference listener as a service, and that's feasible, they could do so and update this code accordingly.

ingomohr commented:
> My 2 cents... Programmatically register to get this PR to closure. Then, if someone wants to work on being able to register a preference listener as a service, and that's feasible, they could do so and update this code accordingly.

Thanks @merks, I'll do that.

ingomohr commented:
#### Change
* Pref change listener is no hacked resource change listener anymore
* Registered the pref change listener at the suggested location when the Workspace opens up
* Deregistered the pref change listener when the Workspace closes
* Performed manual integration test w/ UI fork: still works as expected
* Executed new unit tests and all before-failing unit tests locally: they all still pass (CI will tell, however)
* Tested deregistering pref change listener via debugging. (I'm not sure whether this is supposed to covered w/ a unit test - or if so: which is the best location for that).

laeubi commented:
Please don't register resourcechange listeners in the activator, this will most likely fail if the workspace is not ready yet. You don't need to implement any service when you create a component so something like:

```
@Component(service = {})
public class CheckMissingNaturesListener IPreferenceChangeListener {

	@Reference
	private IWorkspace workspace;

	@Reference
	private ILog log;

	@Reference(target = IScopeContext.BUNDLE_SCOPE_FILTER)
	IScopeContext bundleScope;

	private IEclipsePreferences eclipsePreferences;

	@Activate
	public void register() {
		eclipsePreferences = bundleScope.getNode(""); //$NON-NLS-1$
		eclipsePreferences.addPreferenceChangeListener(this);
	}

	@Deactivate
	public void unregister() {
		eclipsePreferences.removePreferenceChangeListener(this);
	}
}
```

should work here.

merks commented:
It's being done in the Workspace.open not in the activator...   Other things access the INSTANCE scope preferences there already...

laeubi commented:
> It's being done in the Workspace.open not in the activator... Other things access the INSTANCE scope preferences there already...

Yes and too much work is there already done, actually I won't add more things here if not absolutely necessary, registering a resource change listener is not such thing that requires early workspace access...

ingomohr commented:
Thanks for the feedback. It's not a resource change listener, though. It's a prefs change listener. How do you suggest to register it?

ingomohr commented:
Sorry, I missed the snippet above. Will try that approach.

ingomohr commented:
#### Change
* Workspace class is now untouched again
* Pref change listener is registered/deregistered via OSGi service
  * Thanks for the hint! I didn't know this would work, too.
  * Both registration and deregistration were called as expected
* Performed manual integration tests w/ my UI fork: still worked as expected (/)

merks commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ProjectEncodingPreferenceChangeListener.java:
```diff
@@ -0,0 +1,198 @@
+	private static final String MARKER_TYPE = ValidateProjectEncoding.MARKER_TYPE;
+
+	@Reference
+	private IWorkspace workspace;
```
I'm no expert on how this serviced stuff works, but isn't the point here that the workspace value is injected?  I.e., I don't think we need to assign it below.

laeubi replied:
Yes this is correct, also please pass the workspace to the workspace job.

ingomohr replied:
removed the assignment

merks commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ProjectEncodingPreferenceChangeListener.java:
```diff
@@ -0,0 +1,198 @@
+
+	@Activate
+	public void register() {
+		eclipsePreferences = bundleScope.getNode(""); //$NON-NLS-1$
```
Is this the same node as this later?

```
IEclipsePreferences node = InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES);
```

laeubi replied:
The preferences are already available as `private IEclipsePreferences eclipsePreferences;` so no need to do this static stuff here.

merks replied:
> The preferences are already available as `private IEclipsePreferences eclipsePreferences;` so no need to do this static stuff here.

I don't understand this comment.  Where/how is the value assigned?  How is equivalent to the node `InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES)`?

merks replied:
I figured they must be the same node, otherwise the code wouldn't work, but it's certainly not obviously the case.

You mention there is  "no need to call later in static parts" and again, forgive me for being dense, but where exactly do you mean is that other place?  Note that the call to `InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES)` in ValidateProjectEncoding.java which is a different file that does not have access to the field  eclipsePreferences in this preference listener class.   So in which class are you suggesting a change?  Or are you not suggesting a change but only answering the "are the nodes equal" question?

laeubi replied:
> ValidateProjectEncoding.java which is a different file that does not have access to the field eclipsePreferences in this preference listener class.

Okay I missed this then forget about that, still they are the same :-)

> Or are you not suggesting a change but only answering the "are the nodes equal" question?

Yes, even though `ValidateProjectEncoding` might also be converted to a component we should not mix that here then.

> but it's certainly not obviously the case.

This is a quite new feature, the `org.eclipse.core.runtime.preferences.IScopeContext.TYPE_BUNDLE` explains this and if you take a look at `org.eclipse.core.internal.preferences.BundleStateScope` you will see that under the hood it calls `InstanceScope.INSTANCE.getNode(<id of bundle>)` and `ResourcesPlugin.PI_RESOURCES` is a constant to the resource bundle id.

ingomohr replied:
Thanks for the feedback. I'm also for not mixing things up for ``ValidateProjectEncoding`` (at least not for this PR)

merks commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ProjectEncodingPreferenceChangeListener.java:
```diff
@@ -0,0 +1,198 @@
+			final int oldSeverity = event.getOldValue() != null ? Integer.parseInt((String) event.getOldValue())
+					: PreferenceInitializer.PREF_MISSING_ENCODING_MARKER_SEVERITY_DEFAULT;
+
+			workspace = ResourcesPlugin.getWorkspace();
```
I doubt this assignment is needed.

laeubi replied:
Correct, please don't use static access anywhere if you can get stuff injected...

ingomohr replied:
removed that static assignment (line 62). I thought, you meant this assignment in your comment above, as well?

merks commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ProjectEncodingPreferenceChangeListener.java:
```diff
@@ -0,0 +1,198 @@
+		if (markers.isEmpty()) {
+			return;
+		}
+		WorkspaceJob job = new WorkspaceJob(Messages.updateMissingProjectEncodingMarkers) {
```
I suppose we could batch all the projects into a single job.  Though I'm not sure that's actually better...

ingomohr replied:
I might be missing something here, but: Aren't we already batching all the projects into a single job, here? The method is called w/ workspace.root for which all markers are resolved that are updated all w/in the same one job starting at line 77.

ingomohr replied:
Ah, maybe you mean the other job down below on line 121? That one is called for each of the projects. However, I can't tell which option is better in terms of performance or RAM-usage.

merks replied:
Given that changing the preference isn't a common operation I think we shouldn't fuss over the performance...

merks reviewed (commented):
I meant these comments to be all in one review.  Sorry..

merks commented:
> @merks @laeubi : Quick question: From the old Gerrit-workflow I remember that the Eclipse projects were interested in having only one commit per issue. That's why I amended my commit for this PR all the time. Is this still what you want? (Amending commits typically makes re-reviews harder - at least that's my take-away from working w/ Bitbucket Server).

Yes, this is typically what people do.  You can also do a chain of commits, and then squash and merge at the end:

![image](https://user-images.githubusercontent.com/208716/179676904-450d7341-e3a7-48ef-a8d2-cfdb4df80927.png)

merks commented:
@iloveeclipse  Do you have any outstanding concerns?  This looks good to go now...

iloveeclipse commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ProjectEncodingPreferenceChangeListener.java:
```diff
@@ -0,0 +1,196 @@
+			final int oldSeverity = event.getOldValue() != null ? Integer.parseInt((String) event.getOldValue())
+					: PreferenceInitializer.PREF_MISSING_ENCODING_MARKER_SEVERITY_DEFAULT;
+
+			if (isIgnore(newSeverity)) {
```
The if/else with lot of custom code below in this class could be simply replaced to 
```
if (newSeverity != oldSeverity) {
	ValidateProjectEncoding.scheduleWorkspaceValidation((Workspace) workspace);
}
```

And then, looking on what is left (just the prefs listener), the code should be simply moved to `CharsetManager` which is started automatically anyway, so no extra component needed, and it can register this tiny listener in `org.eclipse.core.internal.resources.CharsetManager.startup(IProgressMonitor)` and the remaining code in this method above will do the job.

ingomohr replied:
> tiny listener in org.eclipse.core.internal.resources.CharsetManager.startup(IProgressMonitor)

What about loose coupling via OSGi service and single responsibility (as in SOLID)? Adding the pref listener to the ``ChartSetManager``  would again hardwire that listener to another class - which (as I understood from a former review comment) is not what we want.

iloveeclipse replied:
> > tiny listener in org.eclipse.core.internal.resources.CharsetManager.startup(IProgressMonitor)
> 
> What about loose coupling via OSGi service and single responsibility (as in SOLID)? Adding the pref listener to the `ChartSetManager` would again hardwire that listener to another class - which (as I understood from a former review comment) is not what we want.

Sounds good, but fully missing the point. 

The `ChartSetManager` **is** about managing that stuff for which the preference is added. Why should it violate any rules if exact that manager not only triggers creation of markers but also updates them? With few extra lines of code vs extra "loose coupled" OSGI service that adds a lot more code, complexity and overhead for startup to do exact same thing. That makes no sense.

ingomohr replied:
> Sounds good, but fully missing the point.
>
> The ChartSetManager is about managing that stuff for which the preference is added. Why should it violate any rules if exact that manager not only triggers creation of markers but also updates them? With few extra lines of code vs extra "loose coupled" OSGI service that adds a lot more code, complexity and overhead for startup to do exact same thing. That makes no sense.

@iloveeclipse , @laeubi , @merks 
Ok. I'd say you decide on this together and I'll make the change. I'm not going to make it w/o the consent of at least 2 of you as I already refactored how to connect the pref change listener before and I don't want to invest my free time in refactoring the same code back and forth.

iloveeclipse replied:
> I already refactored how to connect the pref change listener before and I don't want to invest my free time in refactoring the same code back and forth.

There will be no back and forth. There is literally no need in any OSGI service or component here.
There is no need in big refactoring either. All what is needed now to throw away **all** of complex and unnecessary code wiring that extra component which only purpose is to activate and add a small listener and moving that small listener into the CharsetManager where it belongs to.

"If the only tool you have is a hammer, you tend to see every problem as a nail.” 
Here is no nails for OSGI, the code change to add a listener is trivial and doesn't require OSGI.

laeubi replied:
@iloveeclipse This code is neither complex nor a "hammer" it is separation of concerns and all those "WhateverMangers" that are tightly integrated and mangled into the Workspace init with different concerns are better separated instead of "just adding a few lines" her and there and calling back and forth, carefully tweaking order of init calls, keeping care of cleanup manually and so on.

So why not take the opportunity and start right now? Why these reservations and discomfort? All this stuff was added to finally solve the dreaded early Resource-Plugin-Activation problem that was not solved for > 10 years so why not use it and recommend its usage to give examples for other adopters where it is maybe not just a few lines?

@ingomohr I did not requested any changes but was asked how to do it with services and I gave the necessary hints how to do it without static access, but that's not a requirement (from the project) and just to show people now ways instead of "we have always done it that way".

So everyone is free to ignore my suggestions here (that's why I left a comment versus a real review) but then please don't complain that there are "gazillion" of references to the static stuff, or that workspace init sometimes break, is slow or wrecked in any other way :-)

ingomohr replied:
Thank you for the response, @iloveeclipse  and @laeubi.

I'm completely w/ @laeubi on this matter. Still, I refactored the code to drop the service as requested by @iloveeclipse and supported by @merks - simply because we should put an end to this PR. I'm just a contributor and can only decide for my own plug-ins. For those, my take-away is that I can declare those listeners via OSGi to gain all the benefits @laeubi listed. Thank you for that! :)

iloveeclipse reviewed (changes requested):
See my comment on ProjectEncodingPreferenceChangeListener. Lot of unneeded code.

ingomohr commented:
> > @iloveeclipse Do you have any outstanding concerns? This looks good to go now...
> 
> I haven't tested/reviewed, but it looks like the new test isn't added to the test suite and so it will be not executed.

@iloveeclipse You mean ``org.eclipse.core.tests.resources.AllTests``, yes? I would have added the test if I had known the test should be added there.

iloveeclipse commented:
Just look around where other tests are included and add it. I have no IDE right now, but most of Eclipse tests are referenced by some testsuite.

ingomohr commented:
### Change
- Added ``ProjectEncodingTest`` to suite ``org.eclipse.core.tests.resources.AllTests``.

iloveeclipse commented:
> The Jenkins build of this PR has now completed. See details at https://ci.eclipse.org/platform/job/eclipse.platform.resources/job/PR-167/11/

Now you see +10 tests:
https://ci.eclipse.org/platform/job/eclipse.platform.resources/job/PR-167/11/testReport/org.eclipse.core.tests.resources/

Fine. Next, please throw away most listener related code as explained before.

ingomohr commented:
> > The Jenkins build of this PR has now completed. See details at https://ci.eclipse.org/platform/job/eclipse.platform.resources/job/PR-167/11/
> 
> Now you see +10 tests: https://ci.eclipse.org/platform/job/eclipse.platform.resources/job/PR-167/11/testReport/org.eclipse.core.tests.resources/
> 
> Fine. Next, please throw away most listener related code as explained before.

Yup. Seen those test results on Jenkins. DRY violating code is gone.

ingomohr commented:
> @ingomohr: it would be easier to validate this change for reviewers if you could push a PR for platform UI with the preference page change, even if it is not ready yet.

I created [PR 220](https://github.com/eclipse-platform/eclipse.platform.ui/pull/220) for this.

merks commented:
What @iloveeclipse suggested does sound like a simplification.  That being said, I do understand that it's starting to seem like each person that is asked for an opinion pushes in a different direction.  That doesn't make contributing feel very rewarding but more of a gauntlet.  I think we can all recognize that a lot of time has been invested here and if focused feedback had been provided early, the process would have been shorter.

ingomohr commented:
> What @iloveeclipse suggested does sound like a simplification. That being said, I do understand that it's starting to seem like each person that is asked for an opinion pushes in a different direction. That doesn't make contributing feel very rewarding but more of a gauntlet. I think we can all recognize that a lot of time has been invested here and if focused feedback had been provided early, the process would have been shorter.

Thank you for the update/feedback, @merks. That matches my feeling, too.
So, I'll drop the pref-change-listener-via-service approach and add that listener to the ``CharsetManager``. I'll do that tonight.

ingomohr commented:
> Please compare #170 with current code. That does exactly same with less code / less complexity.

Thanks for the hint, @iloveeclipse. The only thing I'm not keen on in your PR is the inner class for the preference change listener. I used a lambda implementation instead b/c it's much smaller.
If you specifically want that inner class to happen, just let me know.

iloveeclipse commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/CharsetManager.java:
```diff
@@ -506,8 +512,21 @@ public void startup(IProgressMonitor monitor) {
 		job = new CharsetManagerJob();
 		resourceChangeListener = new ResourceChangeListener();
 		workspace.addResourceChangeListener(resourceChangeListener, IResourceChangeEvent.POST_CHANGE);
+		initPreferenceChangeListener();
```
Any particular reason why the listener is added before the job that is supposed to create markers? I would move this at very end of the method, just to keep them together.

ingomohr replied:
moved down to the end
(I've semantically-misgrouped the pref change listener w/ the resource change listener - but actually, they apparently form no semantic group)

iloveeclipse commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ValidateProjectEncoding.java:
```diff
@@ -95,6 +99,22 @@ static void updateMissingEncodingMarker(IProject project) {
+	 */
+	private static int getSeverity() {
+		int severity = PreferenceInitializer.PREF_MISSING_ENCODING_MARKER_SEVERITY_DEFAULT;
+		IEclipsePreferences node = InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES);
```
I think this will not work with product customization (where one could set it once for product and never set INSTANCE value). So better for **reading** the value use 

`Platform.getPreferencesService().getInt(ResourcesPlugin.PI_RESOURCES,					PREF_MISSING_ENCODING_MARKER_SEVERITY, severity, null);`

**Writing** goes as before to the instance scope.

ingomohr replied:
replaced reading w/ pref service access

ingomohr replied:
Btw: I had taken that approach from class ``CheckMissingNaturesListener::getMissingNatureSeverity`` (as suggested when I started this issue). Maybe we should create a ticket for fixing that class, too.

iloveeclipse commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/utils/Messages.java:
```diff
@@ -334,6 +335,7 @@ public class Messages extends NLS {
 	public static String WM_mutexAbandoned;
 
 	public static String updateUnknownNatureMarkers;
+	public static String updateMissingProjectEncodingMarkers;
```
This is not used anymore, please remove

ingomohr replied:
removed line (=> no further changes => this class is no longer listed in the PR)

iloveeclipse commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/utils/messages.properties:
```diff
@@ -122,6 +122,7 @@ preferences_saveProblems=Exception occurred while saving project preferences: {0
 preferences_syncException=Exception occurred while synchronizing node: {0}.
 
 updateUnknownNatureMarkers=Update markers for unknown nature on projects
+updateMissingProjectEncodingMarkers="Update markers for missing project encoding
```
This is not used anymore, please remove

ingomohr replied:
removed line (=> no further changes => this class is no longer listed in the PR)

iloveeclipse commented on tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/ProjectEncodingTest.java:
```diff
@@ -0,0 +1,240 @@
+				ResourcesPlugin.PREF_MISSING_ENCODING_MARKER_SEVERITY,
+				PreferenceInitializer.PREF_MISSING_ENCODING_MARKER_SEVERITY_DEFAULT);
+		InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES).flush();
+		super.tearDown();
```
Calling `super.tearDown()` in the middle ow teardown looks suspicious. This should be done at very end.

Also please remove system out below.

ingomohr replied:
Thanks. That was an unfinished left-over. Moved down to end.

iloveeclipse commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java:
```diff
@@ -355,6 +356,14 @@ public final class ResourcesPlugin extends Plugin {
+	 *
+	 * @since 3.18
+	 */
+	public static final String PREF_MISSING_ENCODING_MARKER_SEVERITY = "missingEncodingMarkerSeverity"; //$NON-NLS-1$
```
Could you please add a note about expected / allowed values (especially "ignored" one). It is not clear that other values than `IMarker.SEVERITY_` are also accepted.

ingomohr replied:
Added the info.

Thought: I would be nice to have a severity constant for this like {{IMarker.SEVERITY_IGNORE}}. But I understand that that would require much more discussion as that would affect API.
I "copied" that approach with ``severity < 0`` (again, as suggested when I started this issue) from ``CheckMissingNaturesListener`` (see ``org.eclipse.core.internal.resources.CheckMissingNaturesListener.updateMarkers(Collection<IProject>)``)

iloveeclipse reviewed (changes requested):
OK, the main part seem to be OK now, the rest are smaller issues.

iloveeclipse commented:
@trancexpress : Simeon, as you've originally wrote the code, could you please also look at the PR please? The corresponding part for the UI with new preference is https://github.com/eclipse-platform/eclipse.platform.ui/pull/220.

trancexpress commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ValidateProjectEncoding.java:
```diff
@@ -83,9 +84,10 @@ public IStatus runInWorkspace(IProgressMonitor monitor) {
 				String defaultCharset = getDefaultCharset(project);
-				if (defaultCharset == null) {
-					createEncodingMarker(project);
+				if (severity >= 0 && defaultCharset == null) {
```
```severity != -1``` instead? Since the magic number is -1.

I think you can define the ```-1``` here as a constant and use it in platform UI.

The code in platform UI will complain about using non-API, but I think that is better than having the same magic number hard-coded in 2 places. Even if its unlikely for that magic number to change.

ingomohr replied:
Added constant and used it here.

Updated the UI integration (on the the UI repo) to use that constant too.
(However, that change doesn't feel good to me. Internal code in Eclipse plug-ins is internal to convey the information: "It's safe to alter things here w/o breaking contract for client plugins". But if this is the usual approach between the Eclipse platform bundles anyway, it might be another story. Still, this change wouldn't make things any better imho.)

trancexpress commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ValidateProjectEncoding.java:
```diff
@@ -95,6 +97,19 @@ static void updateMissingEncodingMarker(IProject project) {
+	 * Returns the current severity for
+	 * {@link ResourcesPlugin#PREF_MISSING_ENCODING_MARKER_SEVERITY}.
+	 *
+	 * @return current severity. Default value is nothing else is specified.
```
I don't understand the sentence here, re-wording it would be great.

ingomohr replied:
Thanks. There was a typo involved here. Fixed it.

trancexpress commented on tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/ProjectEncodingTest.java:
```diff
@@ -0,0 +1,238 @@
+		if (project != null) {
+			project.delete(true, null);
+		}
+		InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES).putInt(
```
Wasn't this now done with a preference store (in production code) instead of working on the node directly?

ingomohr replied:
Maybe I'm missing something here, but preference store is an API of JFace. JFace is a UI framework/library and cannot/should not be used in backend plug-ins like this one.
If I'm understanding it correctly, the preference store used in the UI also uses the this instance scope node:

```
missingEncodingSeverityCombo.setPreferenceStore(new ScopedPreferenceStore(InstanceScope.INSTANCE, ResourcesPlugin.PI_RESOURCES));
```

trancexpress replied:
Never mind, I was thinking of ```ResourcesPlugin.getPlugin().getPluginPreferences()```. I checked your code again, you are using the preference service (which is fine) and setting a value with it is not trivial.

trancexpress commented on tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/ProjectEncodingTest.java:
```diff
@@ -0,0 +1,238 @@
+		InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES).putInt(
+				ResourcesPlugin.PREF_MISSING_ENCODING_MARKER_SEVERITY,
+				PreferenceInitializer.PREF_MISSING_ENCODING_MARKER_SEVERITY_DEFAULT);
+		InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES).flush();
```
Don't think this is necessary.

ingomohr replied:
I took-over this like from class ``NatureTest`` (the approach to implement this like for the unknown-nature-check was suggested when I started to work on this issue).

I'd like to keep it as it's a safeguard for tests that run later on because it forces to write the default changes back to the store. (if I'm getting the jdoc right).

trancexpress commented on tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/ProjectEncodingTest.java:
```diff
@@ -0,0 +1,238 @@
+				ResourcesPlugin.PREF_MISSING_ENCODING_MARKER_SEVERITY,
+				PreferenceInitializer.PREF_MISSING_ENCODING_MARKER_SEVERITY_DEFAULT);
+		InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES).flush();
+		getWorkspace().getRoot().refreshLocal(IResource.DEPTH_INFINITE, null);
```
What is that for?

ingomohr replied:
To wait until the workspace job finished. Without waiting the change-prefs tests will fail.

trancexpress replied:
> To wait until the workspace job finished. Without waiting the change-prefs tests will fail.

Can you try either:

```
TestUtil.waitForJobs(getName(), 10, 5_000, null);
```
Or:
```
TestUtil.waitForJobs(getName(), 10, 5_000, ValidateProjectEncoding.class);
```
(if that is the important job here)

ingomohr replied:
Thanks for the hint. Didn't know that class at all. I'll check.

ingomohr replied:
What does that test``TestUtil`` (contained in the test project for the resources plugin) better than the official API from ``JobManager``? I don't see any advantage right now.
* Using the official API is stable throughout 15 PR builds and all local builds w/ I ran before
* I need less parameters
* I don't need some local class that I can't use for testing any Eclipse plugin
* I can use an API that I can consider safely to be fully tested
** I can't find any unit tests for ``TestUtil``.

=> I very much would like to keep using the official job API here.

trancexpress replied:
> What does that test`TestUtil` (contained in the test project for the resources plugin) better than the official API from `JobManager`? I don't see any advantage right now.

I don't understand your question. Currently you are doing a refresh on the workspace (meaning, the test expects some I/O that Eclipse resources API didn't "see", and so a refresh will "sync" what the Eclipse API "sees" and the file system state). You said this is meant to wait for jobs.

Using the test utility to wait on jobs is the standard way of... waiting on jobs during tests. The utility exists since it will do some "nices to have" such as timeouts (necessary for tests that run into hanging jobs) as well as reporting which jobs are still running when the timeout occurs.

But its the same to me. If it starts failing due to whichever workspace job no longer needs a workspace lock, the test will be fixed.

ingomohr replied:
Sorry - I confused this with ``Job.getJobManager().join`` far below. Forget what I said about waiting. I'll check if I ever needed this line in the first place. My bad :)

ingomohr replied:
Yup. Dropped that line. I apparently had copied that line from ``NatureTest`` and didn't re-evaluate whether I need it. Thank you for your feedback on this!

trancexpress commented:
I tested the latest changes here and https://github.com/eclipse-platform/eclipse.platform.ui/pull/220/commits/39b6cd3edb7708d23798a2ee76dbcabbcbd5cf59, LGTM.

The plug-in customization line to disable the marker (I assume the only customization of interest):

```
org.eclipse.core.resources/missingEncodingMarkerSeverity=-1
```

Please mention this in the commit message.

(plug-in customization is specified with command line argument for Eclipse: ```-pluginCustomization /.../plugin_customization.ini```)

ingomohr commented on tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/ProjectEncodingTest.java:
```diff
@@ -0,0 +1,238 @@
+				PreferenceInitializer.PREF_MISSING_ENCODING_MARKER_SEVERITY_DEFAULT);
+		InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES).flush();
+		getWorkspace().getRoot().refreshLocal(IResource.DEPTH_INFINITE, null);
+		ensureDoesNotExistInWorkspace(getWorkspace().getRoot());
```
I didn't write that method nor change any line of code in that class.

From its impl: if the resource exists -> delete it.

=> I'm with you about this method signature not being clear enough. But changing that class is entirely out of scope of my PR. And sure: Modifying some other class might not hurt. But I'm already working on the 16th (yes, 16th) Pull Request build for this feature which is in parts because 4 reviewers had different opions about things (I still appreciate every feedback, though - just to be clear), and I certainly want to extend the discussion to code that I didn't modify myself.

trancexpress replied:
> I didn't write that method nor change any line of code in that class.
> 
> From its impl: if the resource exists -> delete it.

You mean you copied code you don't understand. All good, I don't understand it either. To me it reads like "delete the workspace", which seems odd, but maybe I misunderstand. I see a lot of other tests do the same, so I assume it doesn't hurt.

ingomohr replied:
> You mean you copied code you don't understand

Not at all. _You are referring to a method that is not in my PR but in a class that existed before_. And when I look into that method I understand what it does - and I decided that if (apparently so) that method signature was good enough for other people's calls, then it's good enough for my call, as well.

ingomohr replied:
> To me it reads like "delete the workspace", which seems odd, but maybe I misunderstand

I'm fully w/ you on that one, though. :)

ingomohr replied:
I think I understood you wrong. I first thought your feedback was on the method signature. But perhaps it was more on "why call that method at all".

I deleted that line now. I took it over from ``NatureTest`` (which was initially suggested as template for by change), but we don't need that call at all.

ingomohr commented:
> I tested the latest changes here and [eclipse-platform/eclipse.platform.ui@39b6cd3](https://github.com/eclipse-platform/eclipse.platform.ui/commit/39b6cd3edb7708d23798a2ee76dbcabbcbd5cf59), LGTM.
> 
> The plug-in customization line to disable the marker (I assume the only customization of interest):
> 
> ```
> org.eclipse.core.resources/missingEncodingMarkerSeverity=-1
> ```
> 
> Please mention this in the commit message.
> 
> (plug-in customization is specified with command line argument for Eclipse: `-pluginCustomization /.../plugin_customization.ini`)

I'm not sure I fully understand you here. I never heard of plug-in customization for preferences before. I updated the commit message (I hope) accordingly.

trancexpress commented:
> I'm not sure I fully understand you here. I never heard of plug-in customization for preferences before. I updated the commit message (I hope) accordingly.

Commit message is fine now.

The plug-in customization file (can be passed with the mentioned argument) contains preferences that you would like Eclipse to set on "start-up". The preference values in the file act as defaults and can still be changed by the user. See e.g.: https://www.eclipse.org/eclipse/platform-core/documents/user_settings/plugin_customization.html

ingomohr commented:
> > I'm not sure I fully understand you here. I never heard of plug-in customization for preferences before. I updated the commit message (I hope) accordingly.
> 
> Commit message is fine now.
> 
> The plug-in customization file (can be passed with the mentioned argument) contains preferences that you would like Eclipse to set on "start-up". The preference values in the file act as defaults and can still be changed by the user. See e.g.: https://www.eclipse.org/eclipse/platform-core/documents/user_settings/plugin_customization.html

Thanks for the info! We certainly can use that learning with our Eclipse-based products, too. :)

ingomohr commented:
@trancexpress I think I resolved/answered all your feedback now. (The PR's readability starts to become a little bit complicated ;) ).
This is just the info for you that I don't have any local changes left that I plan to push.

trancexpress commented:
Can be merged from my POV, with the respective change in platform UI.

ingomohr commented:
> Can be merged from my POV, with the respective change in platform UI.

Thanks for your feedback! Appreciate it.

merks commented:
@iloveeclipse  Have all your change requests been addressed?

iloveeclipse commented:
> @iloveeclipse Have all your change requests been addressed?

Yes, thanks.

> But I'm already working on the 16th (yes, 16th) Pull Request build for this feature

So what? I've seen reviews with over 50 rounds or three years. Only the sky is the limit:-)

iloveeclipse reviewed (approved):
LGTM

ingomohr commented:
@merks @laeubi @iloveeclipse @trancexpress 
Thank you all for your feedback and the discussions!

Even though this felt like "quite a ride" (maybe not only to me), I think we added some nice value here.

And maybe future contributions will go a little bit smoother, too. 😊
